### PR TITLE
fix run_model_from_effective_irradiance with iterable weather and excluding cell temperature

### LINKED ISSUE #1163: ValueError: ModelChain.run_from_effective_irradiance([weather]) when only providing temp_air and wind_speed
**Describe the bug**
According to the (new) docstring for `ModelChain.run_from_effective_irradiance`, cell temperature can be calculated from temperature_model using `'effective_irradiance'`. This is not the case when using one or more arrays 
https://github.com/pvlib/pvlib-python/blame/master/pvlib/modelchain.py#L1589-L1606

**To Reproduce**
```python
from copy import deepcopy
import pandas as pd
from pvlib.location import Location
from pvlib.pvsystem import Array, PVSystem
from pvlib.modelchain import ModelChain


array_params = {
    "surface_tilt": 32.0,
    "surface_azimuth": 180.0,
    "module": "Canadian_Solar_Inc__CS5P_220M",
    "albedo": 0.2,
    "temperature_model_parameters": {
        "u_c": 29.0,
        "u_v": 0.0,
        "eta_m": 0.1,
        "alpha_absorption": 0.9,
    },
    "strings": 5,
    "modules_per_string": 7,
    "module_parameters": {
        "alpha_sc": 0.004539,
        "gamma_ref": 1.2,
        "mu_gamma": -0.003,
        "I_L_ref": 5.11426,
        "I_o_ref": 8.10251e-10,
        "R_sh_ref": 381.254,
        "R_sh_0": 400.0,
        "R_s": 1.06602,
        "cells_in_series": 96,
        "R_sh_exp": 5.5,
        "EgRef": 1.121,
    },
}
inverter_parameters = {
    "Paco": 250.0,
    "Pdco": 259.589,
    "Vdco": 40.0,
    "Pso": 2.08961,
    "C0": -4.1e-05,
    "C1": -9.1e-05,
    "C2": 0.000494,
    "C3": -0.013171,
    "Pnt": 0.075,
}


location = Location(latitude=33.98, longitude=-115.323, altitude=2300)

array_sys = PVSystem(
    arrays=[
        Array(**array_params, name=0),
    ],
    inverter_parameters=inverter_parameters,
)
weather = pd.DataFrame(
    {
        "effective_irradiance": [1100.0, 1101.0],
        "temp_air": [25.0, 26.0],
        "wind_speed": [10.0, 10.0],
    },
    index=pd.DatetimeIndex(
        [pd.Timestamp("2021-01-20T12:00-05:00"), pd.Timestamp("2021-01-20T12:05-05:00")]
    ),
)
mc0 = ModelChain(
    array_sys,
    location,
    aoi_model="no_loss",
    spectral_model="no_loss",
)
mc1 = deepcopy(mc0)

mc0.run_model_from_effective_irradiance(weather)
assert isinstance(mc0.results.cell_temperature, pd.Series)


mc1.run_model_from_effective_irradiance([weather])  # ValueError

```

**Expected behavior**
Running the model with both `weather` and `[weather]` work

**Versions:**
 - ``pvlib.__version__``: 0.9.0-alpha.2+5.gb40df75

### PR DESCRIPTION
**Pull request by cwhanse**
- [x] Closes #1163
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries to [`docs/sphinx/source/api.rst`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/api.rst) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

Moves check for complete POA for cell temperature calculations to `ModelChain._set_celltemp`.
Adds testing for weather/data input type (tuple, list) to tests for `ModelChain.run_from_effective_irradiance`

### CODE REVIEW

**Comment by cwhanse** [comment]
@wholmgren @wfvining one [test ](https://dev.azure.com/solararbiter/pvlib%20python/_build/results?buildId=4807&view=logs&j=83ed75f2-3f6e-51e0-dc7b-dfceecefbbcd&t=0b9cc177-13eb-5a9e-870b-984e27ba97f8&l=156) is failing. I actually think the ModelChain should run with that that test setup, since `effective_irradiance` is provided for every array. The cell temperature model can use `effective_irradiance` since #1129. PTAL and let me know if you agree.

**Comment by wfvining** [comment]
That does seem reasonable to me. If we remove that test, we should make sure that there is still a test that covers the case where one array does not have enough information to calculate cell temperature.

**Comment by wholmgren** [comment]
I agree.

**Comment by cwhanse** [comment]
> That does seem reasonable to me. If we remove that test, we should make sure that there is still a test that covers the case where one array does not have enough information to calculate cell temperature.

I had to call `_set_celltemp` directly to get a case that raises the `ValueError`. Missing `'poa_global'` is covered by the new `test__set_celltemp_missing_poa`. Missing `'poa_global'` or `effective_irradiance` are caught upstream by the `_check_multiple_input` steps of `run_from_poa` and `run_from_effective_irradiance`.

**Comment by wholmgren** [comment]
> I had to call _set_celltemp directly to get a case that raises the ValueError

This suggests to me that the ValueError is only accessible from private calls and cannot be hit through `run_*` method calls. I am I misunderstanding?

**Comment by cwhanse** [comment]
> > I had to call _set_celltemp directly to get a case that raises the ValueError
> 
> This suggests to me that the ValueError is only accessible from private calls and cannot be hit through `run_*` method calls. Am I misunderstanding?

I believe that to be the case, but could be wrong.

**Comment by wholmgren** [comment]
Why keep it if we can't hit it through the run method calls?

**Comment by cwhanse** [comment]
> > > I had to call _set_celltemp directly to get a case that raises the ValueError
> > 
> > 
> > This suggests to me that the ValueError is only accessible from private calls and cannot be hit through `run_*` method calls. Am I misunderstanding?

- With `ModelChain.run_model`, `prepare_inputs` calls `_check_multiple_input` to verify that `weather` has the same dimension as num_arrays, and calls  `_verify_df` to verify that 'ghi', 'dhi' and 'dni' are present in `weather`.
- With `ModelChain.run_model_from_poa`, `prepare_inputs_from_poa` calls `_check_multiple_input` to verify that `data` has the same dimension as num_arrays, and `_verify_df` to verify that `data` contains `'poa_global'`, `'poa_direct'` and `'poa_diffuse'`.
- With `ModelChain.run_model_from_effective_irradiance`, there is no `prepare_inputs` method. `_check_multiple_input` is called to verify that `data` has the same dimension as num_arrays. **There is no use of `_verify_df` to check for `effective_irradiance` in `data`.**  Seems like an oversight here.

All three public `run_model` methods end up at `_run_from_effective_irrad`, which calls `_prepare_temperature` and that's how execution gets to `_set_celltemp`.

**Comment by cwhanse** [comment]
> Why keep it if we can't hit it through the run method calls?

Good question. I'm OK removing that check. @wfvining ?

**Comment by wholmgren** [comment]
> There is no use of _verify_df to check for effective_irradiance in data. Seems like an oversight here.

I agree.

Maybe out of scope for this PR, but I think we could refactor `_prepare_temperature` to use `_verify_df` in try/except blocks that parallel the logic in `run_model_from_effective_irradiance` docstring. I still struggle to understand `_prepare_temperature` every time I come back to it.

**Comment by wholmgren** [comment]
We also need to add or change a test to cover the specific case in #1163. All of the `test_run_model_from_effective_irradiance_*` functions include POA data, while #1163 only has effective irradiance.

**Comment by wholmgren** [comment]
#1163 is `test_run_model_from_effective_irradiance_no_poa_global` but with an iterable input.

**Comment by wfvining** [comment]
> 
> 
> > Why keep it if we can't hit it through the run method calls?
> 
> Good question. I'm OK removing that check. @wfvining ?

I'm okay removing it or just making it catch any generic exception as opposed to a specific `ValueError`. I think given the amount of state maintained inside `ModelChain`, once missing data is allowed inside of modelchain it can be very difficult to notice or debug problems, so these kinds of tests are pretty valuable.

> Maybe out of scope for this PR, but I think we could refactor _prepare_temperature to use _verify_df in try/except blocks that parallel the logic in run_model_from_effective_irradiance docstring. I still struggle to understand _prepare_temperature every time I come back to it

I would support refactoring this. My guess is that the bulk of the technical debt in `ModelChain` is related to `_prepare_temperature`.

**Comment by wholmgren** [comment]
I think this should do it:

```python
@pytest.mark.parametrize("input_type", [lambda x: x[0], tuple, list])
def test_run_model_from_effective_irradiance_no_poa_global(
        sapm_dc_snl_ac_system, location, weather, total_irrad, input_type):
    data = weather.copy()
    data['effective_irradiance'] = total_irrad['poa_global']
    mc = ModelChain(sapm_dc_snl_ac_system, location, aoi_model='no_loss',
                    spectral_model='no_loss')
    ac = mc.run_model_from_effective_irradiance(input_type((data,))).results.ac
    expected = pd.Series(np.array([149.280238, 96.678385]),
                         index=data.index)
    if isinstance(ac, (tuple, list)):
        ac = ac[0]
    assert_series_equal(ac, expected)
```

**Comment by cwhanse** [comment]
> I would support refactoring this. 

As would I. New issue though.

I would also propose that we consider combining `_check_multiple_inputs` and `_verify_df` to hopefully make these checks more easily understood in the future. The only separate use of this pair of methods is in `complete_irradiance` which doesn't use `_verify_df` and instead explicitly looks for 2 of 'ghi', 'dhi' and 'dni'. Maybe that check could be built into `_verify_df`.

File: pvlib/tests/test_modelchain.py [hunk 1]
```diff
@@ -985,26 +985,46 @@ def test_run_model_from_poa_tracking(sapm_dc_snl_ac_system, location,
...
     data['effective_irradiance'] = data['poa_global']
     mc = ModelChain(sapm_dc_snl_ac_system, location, aoi_model='no_loss',
                     spectral_model='no_loss')
-    ac = mc.run_model_from_effective_irradiance(data).results.ac
+    ac = mc.run_model_from_effective_irradiance(input_type((data,))).results.ac
     expected = pd.Series(np.array([149.280238, 96.678385]),
                          index=data.index)
     assert_series_equal(ac, expected)
+    # check with data as an iterable of length 1
+    mc.run_model_from_effective_irradiance(input_type((data,)))
```
Comment (wholmgren) [inline pvlib/tests/test_modelchain.py:1001]:
```suggestion
```

File: pvlib/tests/test_modelchain.py [hunk 2]
```diff
@@ -985,26 +985,46 @@ def test_run_model_from_poa_tracking(sapm_dc_snl_ac_system, location,
...
+
+
+@pytest.mark.parametrize("input_type", [tuple, list])
+def test_run_model_from_effective_irradiance_input_type(
```
Comment (wholmgren) [inline pvlib/tests/test_modelchain.py:1005]:
```suggestion
def test_run_model_from_effective_irradiance_multi_array(
```

File: pvlib/modelchain.py [hunk 3]
```diff
@@ -1588,22 +1590,7 @@ def _prepare_temperature(self, data=None):
...
-            # run_model_from_effective_irradiance() can get to this point
-            # without known POA we can suggest a very specific remedy in the
-            # error message.
-            raise ValueError("Incomplete input data. Data must contain "
-                             "'poa_global'. For systems with multiple Arrays "
-                             "if you have provided 'cell_temperature' for "
-                             "only a subset of Arrays you must provide "
-                             "'poa_global' for all Arrays, including those "
-                             "that have a known 'cell_temperature'.")
+        # ModelChain.temperature_model().
```
Comment (wholmgren) [inline pvlib/modelchain.py:1593]:
```suggestion
```

**Review by wholmgren — APPROVED** [review]
minor suggestions below. I confirmed this resolves the example in #1163.

**Comment by cwhanse** [comment]
Ok to merge

**Pull request merged**